sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs

```diff
@@ -0,0 +1,188 @@
+        /// <param name="consumerGroup">The name of the consumer group this partition pump is associated with.  Events are read in the context of this group.</param>
+        /// <param name="partitionId">The identifier of the Event Hub partition this partition pump is associated with.  Events will be read only from this partition.</param>
+        /// <param name="partitionProcessor">A partition processor used to process events and errors.  Its implementation must be provided by the caller.</param>
+        /// <param name="options">The set of options to use for this partition pump.</param>
+        ///
+        internal PartitionPump(EventHubClient eventHubClient,
+                               string consumerGroup,
+                               string partitionId,
+                               IPartitionProcessor partitionProcessor,
+                               EventProcessorOptions options)
+        {
+            InnerClient = eventHubClient;
+            ConsumerGroup = consumerGroup;
+            PartitionId = partitionId;
+            PartitionProcessor = partitionProcessor;
+            Options = options;
+        }
+
+        /// <summary>
+        ///   Starts the partition pump.  In case it's already running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task Start()
+        {
+            if (RunningTask == null)
+            {
+                RunningTaskTokenSource = new CancellationTokenSource();
+
+                InnerConsumer = InnerClient.CreateConsumer(ConsumerGroup, PartitionId, Options.InitialEventPosition);
+
+                await PartitionProcessor.Initialize().ConfigureAwait(false);
+
+                RunningTask = Run(RunningTaskTokenSource.Token);
+            }
+        }
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task Stop() => await Stop(PartitionProcessorCloseReason.Shutdown);
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <param name="reason">The reason why the partition pump is being closed.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task Stop(PartitionProcessorCloseReason reason)
+        {
+            if (RunningTask != null)
+            {
+                RunningTaskTokenSource.Cancel();
+                RunningTaskTokenSource = null;
+
+                await RunningTask;
+                RunningTask = null;
+
+                await InnerConsumer.CloseAsync();
```

Comment:
This should have a `ConfigureAwait(false)` on there, as should the `RunningTask` above.   Basically, all `await` calls.